src/validate.ts

```diff
@@ -74,6 +82,49 @@ export function validate(
     return [];
   }
 
+  function validateFunctionNode(node: FunctionInterface) {
+    const func =
+      node.arguments.length === 4 ? node.arguments[3] : node.arguments[2];
+
+    return [
+      // visit all other arguments
+      ...node.arguments
+        .filter((arg) => arg !== func)
+        .flatMap((arg) => validateEachChildRecursive(arg, visit)),
+      // process the function closure
+      ...validateEachChildRecursive(func, validateFunctionClosureNode),
+    ];
+  }
+
+  function validateFunctionClosureNode(
+    node: typescript.Node
+  ): typescript.Diagnostic[] {
+    if (
+      checker.isStepFunction(node) ||
+      checker.isTable(node) ||
+      checker.isFunctionlessFunction(node)
+    ) {
+      if (
+        typescript.isPropertyAccessExpression(node.parent) &&
+        node.parent.name.text === "resource"
+      ) {
+        return [
+          newError(node, ErrorCodes.Cannot_Use_Infra_Resource_In_Function),
+        ];
+      }
+    } else if (checker.isEventBus(node)) {
+      if (
+        typescript.isPropertyAccessExpression(node.parent) &&
+        node.parent.name.text === "bus"
```

Comment:
Let's make the property names consistent.

Reply:
done